metaflow/plugins/aip/argo_utils.py

```diff
@@ -14,16 +14,38 @@
 
 def run_argo_workflow(
     kubernetes_namespace: str,
-    template_name: str,
+    template_name: Optional[str] = None,
+    project_name: Optional[str] = None,
+    branch_name: Optional[str] = None,
+    template_name_prefix: Optional[str] = None,
```

Comment:
To make migration easier and follow the same pattern as WFSDK v1 in https://github.com/zillow/metaflow/blob/16459a7a7615d096b117928a433d0c12ba81e818/metaflow/plugins/kfp/kfp_utils.py#L103-L109

How about 

```python
def run_argo_workflow(
    kubernetes_namespace: str,
    template_name: Optional[str] = None,
    template_name_full_version: Optional[str] = None,
    project_name: Optional[str] = None,
    branch_name: Optional[str] = None,
```

Where like in v1 `template_name` runs the latest and in this case is the Workflow name from the users perspective just like `pipeline_name`, from AIP perspective if it is a prefix search (an internal detail of how we get the latest).

And if the user opts, they can set `template_name_full_version` for an exact version trigger.  One of the arguments must be set, but not both.

This follows the same pattern as in v1, and is user centric.

Reply:
For context on what are customers are using, I [searched GitLab](https://gitlab.zgtools.net/search?group_id=181&scope=blobs&search=run_kubeflow_pipeline) for customer projects using the function run_kubeflow_pipeline() in projects. Almost all used the parameter [pipeline_name](https://gitlab.zgtools.net/search?group_id=181&scope=blobs&search=run_kubeflow_pipeline+pipeline_name). None used [pipeline_id](https://gitlab.zgtools.net/search?group_id=181&scope=blobs&search=run_kubeflow_pipeline+pipeline_id) or [pipeline_version_id](https://gitlab.zgtools.net/search?group_id=181&scope=blobs&search=run_kubeflow_pipeline+pipeline_version_id).

It is possible that customers are using pipeline_id and/or pipeline_version_id in Notebooks and would not know it. The above does indicate we should focus on making sure that our replacement for pipeline_name works similarly to the v1 functionality.

Reply:
I agree with Taleb's point that template_name should operate as a prefix search from the AIP perspective. We should have a comment in the code to clarify why it is called template_name even though the AIP code will treat it like a template prefix.

Mapping out the old params to new params (based of Taleb's comment), we'd have:

```
    kubeflow_namespace: str,                   -> kubernetes_namespace: str,
    pipeline_name: Optional[str] = None,       -> template_name: Optional[str] = None,
    pipeline_id: Optional[str] = None,         -> [removed because this concept doesn't exist in Argo]
    pipeline_version_id: Optional[str] = None, -> template_name_full_version: Optional[str] = None,
```

I am not sure how to map v1's `triggered_run_name: Optional[str]` and  `experiment_name: Optional[str]` to v2's `project_name: Optional[str]` and `branch_name: Optional[str]`.

Would `experiment_name: Optional[str]` map to `project_name: Optional[str]`?

Reply:
If we go with Taleb's proposed params, would it make sense to replace `template_name_full_version` with `template_name_specific_version` to improve clarity?

Reply:
Thankfully I do not see any customer using `run_argo_workflow` yet through a GitLab code search, so we're not too concerned with breaking changes here. I'll align with the older version.

Reply:
I think we avoid `template_name` as it can be confusing. Users either use `template_name_prefix` (with other filters) or `template_name_exact` alone.

Reply:
@cavandervoort  There is no more experiment name as the concept is entirely gone. Please let me know if `experiment_name` is used by any customer teams. 

`project_name` is the GitLab project name, which is captured as a workflow template label in v4 CICD.

Reply:
Marking this thread as resolved - please check my new approaches. Specifically I added the `filter_func`, which grants self service potential to customers if they desire unimplemented filtering abilities.

Reply:
Three [projects](https://gitlab.zgtools.net/search?group_id=181&scope=blobs&search=run_kubeflow_pipeline+experiment_name) from Lego Metrics use `experiment_name`.